File: media-attachments/php/media_attachments.php
<?php
/**
 * SendSeven API - Media Attachments Example (PHP)
 *
 * Demonstrates how to upload files and send media messages using the SendSeven API.
 */

// Load environment variables from .env file (simple implementation)
function loadEnv(string $path): void {
    if (!file_exists($path)) {
        return;
    }
    $lines = file($path, FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES);
    foreach ($lines as $line) {
        if (strpos(trim($line), '#') === 0) {
            continue;
        }
        if (strpos($line, '=') !== false) {
            list($name, $value) = explode('=', $line, 2);
            putenv(trim($name) . '=' . trim($value));
        }
    }
}

loadEnv(__DIR__ . '/.env');

// Configuration from environment
$API_TOKEN = getenv('SENDSEVEN_API_TOKEN');
$TENANT_ID = getenv('SENDSEVEN_TENANT_ID');
$API_URL = getenv('SENDSEVEN_API_URL') ?: 'https://api.sendseven.com/api/v1';
$CONVERSATION_ID = getenv('CONVERSATION_ID');

// File size limits (in bytes)
define('IMAGE_MAX_SIZE', 16 * 1024 * 1024);      // 16 MB
define('DOCUMENT_MAX_SIZE', 100 * 1024 * 1024);  // 100 MB
define('VIDEO_MAX_SIZE', 16 * 1024 * 1024);      // 16 MB
define('AUDIO_MAX_SIZE', 16 * 1024 * 1024);      // 16 MB

// Supported content types by message type
$SUPPORTED_TYPES = [
    'image' => ['image/jpeg', 'image/png', 'image/gif', 'image/webp'],
    'document' => [
        'application/pdf',
        'application/msword',
        'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
        'application/vnd.ms-excel',
        'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
        'application/vnd.ms-powerpoint',
        'application/vnd.openxmlformats-officedocument.presentationml.presentation',
        'text/plain',
    ],
    'video' => ['video/mp4', 'video/3gpp'],
    'audio' => ['audio/aac', 'audio/mpeg', 'audio/ogg', 'audio/amr', 'audio/opus'],
];

// Extension to content type mapping
$CONTENT_TYPES = [
    // Images
    'jpg' => 'image/jpeg',
    'jpeg' => 'image/jpeg',
    'png' => 'image/png',
    'gif' => 'image/gif',
    'webp' => 'image/webp',
    // Documents
    'pdf' => 'application/pdf',
    'doc' => 'application/msword',
    'docx' => 'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
    'xls' => 'application/vnd.ms-excel',
    'xlsx' => 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
    'ppt' => 'application/vnd.ms-powerpoint',
    'pptx' => 'application/vnd.openxmlformats-officedocument.presentationml.presentation',
    'txt' => 'text/plain',
    // Video
    'mp4' => 'video/mp4',
    '3gp' => 'video/3gpp',
    // Audio
    'aac' => 'audio/aac',
    'mp3' => 'audio/mpeg',
    'ogg' => 'audio/ogg',
    'amr' => 'audio/amr',
    'opus' => 'audio/opus',
];

/**
 * Get content type from file extension.
 *
 * @param string $filePath Path to the file
 * @return string The content type
 */
function getContentType(string $filePath): string {
    global $CONTENT_TYPES;
    $ext = strtolower(pathinfo($filePath, PATHINFO_EXTENSION));
    return $CONTENT_TYPES[$ext] ?? 'application/octet-stream';
}

/**
 * Get message type from content type.
 *
 * @param string $contentType The content type
 * @return string The message type (image, document, video, audio)
 * @throws Exception If content type is unsupported
 */
function getMessageType(string $contentType): string {
    global $SUPPORTED_TYPES;
    foreach ($SUPPORTED_TYPES as $msgType => $types) {
        if (in_array($contentType, $types)) {
            return $msgType;
        }
    }
    throw new Exception("Unsupported content type: $contentType");
}

/**
 * Get maximum file size for a message type.
 *
 * @param string $messageType The message type
 * @return int Maximum size in bytes
 */
function getMaxSize(string $messageType): int {
    $limits = [
        'image' => IMAGE_MAX_SIZE,
        'document' => DOCUMENT_MAX_SIZE,
        'video' => VIDEO_MAX_SIZE,
        'audio' => AUDIO_MAX_SIZE,
    ];
    return $limits[$messageType] ?? DOCUMENT_MAX_SIZE;
}

/**
 * Upload a file as an attachment.
 *
 * @param string $filePath Path to the file to upload
 * @return array The created attachment object
 * @throws Exception If the API request fails
 */
function uploadAttachment(string $filePath): array {
    global $API_TOKEN, $TENANT_ID, $API_URL;

    // Validate file exists
    if (!file_exists($filePath)) {
        throw new Exception("File not found: $filePath");
    }

    $fileSize = filesize($filePath);
    $filename = basename($filePath);
    $contentType = getContentType($filePath);
    $messageType = getMessageType($contentType);

    // Check file size
    $maxSize = getMaxSize($messageType);
    if ($fileSize > $maxSize) {
        throw new Exception("File too large: $fileSize bytes (max $maxSize bytes for $messageType)");
    }

    $url = $API_URL . '/attachments';

    // Create cURL file for upload
    $cfile = new CURLFile($filePath, $contentType, $filename);

    $ch = curl_init($url);
    curl_setopt_array($ch, [
        CURLOPT_POST => true,
        CURLOPT_POSTFIELDS => ['file' => $cfile],
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_HTTPHEADER => [
            'Authorization: Bearer ' . $API_TOKEN,
            'X-Tenant-ID: ' . $TENANT_ID,
        ],
    ]);

    $response = curl_exec($ch);
    $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
    $error = curl_error($ch);
    curl_close($ch);

    if ($error) {
        throw new Exception("cURL Error: $error");
    }

    if ($httpCode === 413) {
        throw new Exception("File too large (server rejected)");
    } elseif ($httpCode === 415) {
        throw new Exception("Unsupported media type (server rejected)");
    } elseif ($httpCode >= 400) {
        throw new Exception("API Error $httpCode: $response");
    }

    return json_decode($response, true);
}

/**
 * Send a message with an attachment.
 *
 * @param string $conversationId The UUID of the conversation
 * @param string $attachmentId The UUID of the uploaded attachment
 * @param string $messageType Type of message (image, document, video, audio)
 * @param string|null $caption Optional text caption for the message
 * @return array The created message object
 * @throws Exception If the API request fails
 */
function sendMediaMessage(string $conversationId, string $attachmentId, string $messageType, ?string $caption = null): array {
    global $API_TOKEN, $TENANT_ID, $API_URL;

    $url = $API_URL . '/messages';

    $payload = [
        'conversation_id' => $conversationId,
        'message_type' => $messageType,
        'attachments' => [$attachmentId],
    ];

    if ($caption) {
        $payload['text'] = $caption;
    }

    $ch = curl_init($url);
    curl_setopt_array($ch, [
        CURLOPT_POST => true,
        CURLOPT_POSTFIELDS => json_encode($payload),
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_HTTPHEADER => [
            'Authorization: Bearer ' . $API_TOKEN,
            'X-Tenant-ID: ' . $TENANT_ID,
            'Content-Type: application/json',
        ],
    ]);

    $response = curl_exec($ch);
    $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
    $error = curl_error($ch);
    curl_close($ch);

    if ($error) {
        throw new Exception("cURL Error: $error");
    }

    if ($httpCode >= 400) {
        throw new Exception("API Error $httpCode: $response");
    }

    return json_decode($response, true);
}

/**
 * Download an attachment by ID.
 *
 * @param string $attachmentId The UUID of the attachment
 * @param string $outputPath Path to save the downloaded file
 * @throws Exception If the API request fails
 */
function downloadAttachment(string $attachmentId, string $outputPath): void {
    global $API_TOKEN, $TENANT_ID, $API_URL;

    $url = $API_URL . '/attachments/' . $attachmentId . '/download';

    $ch = curl_init($url);
    curl_setopt_array($ch, [
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_HTTPHEADER => [
            'Authorization: Bearer ' . $API_TOKEN,
            'X-Tenant-ID: ' . $TENANT_ID,
        ],
    ]);

    $response = curl_exec($ch);
    $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
    $error = curl_error($ch);
    curl_close($ch);

    if ($error) {
        throw new Exception("cURL Error: $error");
    }

    if ($httpCode >= 400) {
        throw new Exception("API Error $httpCode: $response");
    }

    file_put_contents($outputPath, $response);
    echo "Downloaded to: $outputPath\n";
}

/**
 * Upload and send an image message.
 *
 * @param string $conversationId The UUID of the conversation
 * @param string $filePath Path to the image file
 * @param string|null $caption Optional caption for the image
 * @return array The created message object
 */
function sendImage(string $conversationId, string $filePath, ?string $caption = null): array {
    echo "Uploading image: $filePath\n";
    $attachment = uploadAttachment($filePath);
    echo "  Uploaded: {$attachment['id']}\n";

    echo "Sending image message...\n";
    return sendMediaMessage($conversationId, $attachment['id'], 'image', $caption);
}

/**
 * Upload and send a document message.
 *
 * @param string $conversationId The UUID of the conversation
 * @param string $filePath Path to the document file
 * @param string|null $caption Optional caption for the document
 * @return array The created message object
 */
function sendDocument(string $conversationId, string $filePath, ?string $caption = null): array {
    echo "Uploading document: $filePath\n";
    $attachment = uploadAttachment($filePath);
    echo "  Uploaded: {$attachment['id']}\n";

    echo "Sending document message...\n";
    return sendMediaMessage($conversationId, $attachment['id'], 'document', $caption);
}

/**
 * Upload and send a video message.
 *
 * @param string $conversationId The UUID of the conversation
 * @param string $filePath Path to the video file
 * @param string|null $caption Optional caption for the video
 * @return array The created message object
 */
function sendVideo(string $conversationId, string $filePath, ?string $caption = null): array {
    echo "Uploading video: $filePath\n";
    $attachment = uploadAttachment($filePath);
    echo "  Uploaded: {$attachment['id']}\n";

    echo "Sending video message...\n";
    return sendMediaMessage($conversationId, $attachment['id'], 'video', $caption);
}

/**
 * Upload and send an audio message.
 *
 * @param string $conversationId The UUID of the conversation
 * @param string $filePath Path to the audio file
 * @param string|null $caption Optional caption for the audio
 * @return array The created message object
 */
function sendAudio(string $conversationId, string $filePath, ?string $caption = null): array {
    echo "Uploading audio: $filePath\n";
    $attachment = uploadAttachment($filePath);
    echo "  Uploaded: {$attachment['id']}\n";

    echo "Sending audio message...\n";
    return sendMediaMessage($conversationId, $attachment['id'], 'audio', $caption);
}

/**
 * Demo: Upload a file and send it as a message.
 * Automatically detects the appropriate message type.
 */
function demoUploadAndSend(string $filePath): array {
    global $CONVERSATION_ID;

    $contentType = getContentType($filePath);
    $messageType = getMessageType($contentType);

    echo "\n--- Sending $messageType ---\n";
    echo "File: $filePath\n";
    echo "Content-Type: $contentType\n";

    $attachment = uploadAttachment($filePath);
    echo "Attachment uploaded:\n";
    echo "  ID: {$attachment['id']}\n";
    echo "  Filename: {$attachment['filename']}\n";
    echo "  Size: {$attachment['file_size']} bytes\n";

    $message = sendMediaMessage(
        $CONVERSATION_ID,
        $attachment['id'],
        $messageType,
        "Here's a $messageType file!"
    );

    echo "Message sent:\n";
    echo "  ID: {$message['id']}\n";
    echo "  Status: {$message['status']}\n";
    echo "  Created at: {$message['created_at']}\n";

    return $message;
}

// Main execution
function main(): void {
    global $API_TOKEN, $TENANT_ID, $API_URL, $CONVERSATION_ID, $argv;

    // Validate configuration
    if (!$API_TOKEN) {
        echo "Error: SENDSEVEN_API_TOKEN environment variable is required\n";
        exit(1);
    }

    if (!$TENANT_ID) {
        echo "Error: SENDSEVEN_TENANT_ID environment variable is required\n";
        exit(1);
    }

    if (!$CONVERSATION_ID) {
        echo "Error: CONVERSATION_ID environment variable is required\n";
        exit(1);
    }

    echo "SendSeven Media Attachments Example\n";
    echo str_repeat('=', 40) . "\n";
    echo "API URL: $API_URL\n";
    echo "Conversation: $CONVERSATION_ID\n";

    // Check for command line argument (file to upload)
    $filePath = $argv[1] ?? null;

    if ($filePath) {
        try {
            demoUploadAndSend($filePath);
        } catch (Exception $e) {
            echo "Error: " . $e->getMessage() . "\n";
            exit(1);
        }
    } else {
        echo "\nUsage: php media_attachments.php <file_path>\n";
        echo "\nSupported file types:\n";
        echo "  Images:    .jpg, .jpeg, .png, .gif, .webp (max 16 MB)\n";
        echo "  Documents: .pdf, .doc, .docx, .xls, .xlsx, .ppt, .pptx, .txt (max 100 MB)\n";
        echo "  Video:     .mp4, .3gp (max 16 MB)\n";
        echo "  Audio:     .aac, .mp3, .ogg, .amr, .opus (max 16 MB)\n";
        echo "\nExample:\n";
        echo "  php media_attachments.php /path/to/image.jpg\n";

        // Demo with a sample file if it exists
        $sampleFiles = ['sample.jpg', 'sample.png', 'sample.pdf'];
        foreach ($sampleFiles as $sample) {
            if (file_exists($sample)) {
                echo "\nFound sample file: $sample\n";
                try {
                    demoUploadAndSend($sample);
                } catch (Exception $e) {
                    echo "Error: " . $e->getMessage() . "\n";
                }
                break;
            }
        }
    }
}

main();

File: media-attachments/csharp/MediaAttachments.cs
/**
 * SendSeven API - Media Attachments Example (C#)
 *
 * Demonstrates how to upload files and send media messages using the SendSeven API.
 */

using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;

class Program
{
    private static readonly HttpClient httpClient = new HttpClient();

    // Configuration from environment
    private static string ApiToken => Environment.GetEnvironmentVariable("SENDSEVEN_API_TOKEN") ?? "";
    private static string TenantId => Environment.GetEnvironmentVariable("SENDSEVEN_TENANT_ID") ?? "";
    private static string ApiUrl => Environment.GetEnvironmentVariable("SENDSEVEN_API_URL") ?? "https://api.sendseven.com/api/v1";
    private static string ConversationId => Environment.GetEnvironmentVariable("CONVERSATION_ID") ?? "";

    // File size limits (in bytes)
    private const long ImageMaxSize = 16L * 1024 * 1024;      // 16 MB
    private const long DocumentMaxSize = 100L * 1024 * 1024;  // 100 MB
    private const long VideoMaxSize = 16L * 1024 * 1024;      // 16 MB
    private const long AudioMaxSize = 16L * 1024 * 1024;      // 16 MB

    // Supported content types by message type
    private static readonly Dictionary<string, string[]> SupportedTypes = new()
    {
        ["image"] = new[] { "image/jpeg", "image/png", "image/gif", "image/webp" },
        ["document"] = new[] {
            "application/pdf", "application/msword",
            "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
            "application/vnd.ms-excel",
            "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
            "application/vnd.ms-powerpoint",
            "application/vnd.openxmlformats-officedocument.presentationml.presentation",
            "text/plain"
        },
        ["video"] = new[] { "video/mp4", "video/3gpp" },
        ["audio"] = new[] { "audio/aac", "audio/mpeg", "audio/ogg", "audio/amr", "audio/opus" }
    };

    // Extension to content type mapping
    private static readonly Dictionary<string, string> ContentTypes = new()
    {
        // Images
        [".jpg"] = "image/jpeg",
        [".jpeg"] = "image/jpeg",
        [".png"] = "image/png",
        [".gif"] = "image/gif",
        [".webp"] = "image/webp",
        // Documents
        [".pdf"] = "application/pdf",
        [".doc"] = "application/msword",
        [".docx"] = "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
        [".xls"] = "application/vnd.ms-excel",
        [".xlsx"] = "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        [".ppt"] = "application/vnd.ms-powerpoint",
        [".pptx"] = "application/vnd.openxmlformats-officedocument.presentationml.presentation",
        [".txt"] = "text/plain",
        // Video
        [".mp4"] = "video/mp4",
        [".3gp"] = "video/3gpp",
        // Audio
        [".aac"] = "audio/aac",
        [".mp3"] = "audio/mpeg",
        [".ogg"] = "audio/ogg",
        [".amr"] = "audio/amr",
        [".opus"] = "audio/opus"
    };

    static async Task Main(string[] args)
    {
        // Load .env file if it exists
        LoadEnvFile();

        // Validate configuration
        if (string.IsNullOrEmpty(ApiToken))
        {
            Console.WriteLine("Error: SENDSEVEN_API_TOKEN environment variable is required");
            Environment.Exit(1);
        }

        if (string.IsNullOrEmpty(TenantId))
        {
            Console.WriteLine("Error: SENDSEVEN_TENANT_ID environment variable is required");
            Environment.Exit(1);
        }

        if (string.IsNullOrEmpty(ConversationId))
        {
            Console.WriteLine("Error: CONVERSATION_ID environment variable is required");
            Environment.Exit(1);
        }

        Console.WriteLine("SendSeven Media Attachments Example");
        Console.WriteLine(new string('=', 40));
        Console.WriteLine($"API URL: {ApiUrl}");
        Console.WriteLine($"Conversation: {ConversationId}");

        // Check for command line argument (file to upload)
        if (args.Length > 0)
        {
            try
            {
                await DemoUploadAndSendAsync(args[0]);
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error: {ex.Message}");
                Environment.Exit(1);
            }
        }
        else
        {
            Console.WriteLine("\nUsage: dotnet run <file_path>");
            Console.WriteLine("\nSupported file types:");
            Console.WriteLine("  Images:    .jpg, .jpeg, .png, .gif, .webp (max 16 MB)");
            Console.WriteLine("  Documents: .pdf, .doc, .docx, .xls, .xlsx, .ppt, .pptx, .txt (max 100 MB)");
            Console.WriteLine("  Video:     .mp4, .3gp (max 16 MB)");
            Console.WriteLine("  Audio:     .aac, .mp3, .ogg, .amr, .opus (max 16 MB)");
            Console.WriteLine("\nExample:");
            Console.WriteLine("  dotnet run /path/to/image.jpg");

            // Demo with a sample file if it exists
            string[] sampleFiles = { "sample.jpg", "sample.png", "sample.pdf" };
            foreach (var sample in sampleFiles)
            {
                if (File.Exists(sample))
                {
                    Console.WriteLine($"\nFound sample file: {sample}");
                    try
                    {
                        await DemoUploadAndSendAsync(sample);
                    }
                    catch (Exception ex)
                    {
                        Console.WriteLine($"Error: {ex.Message}");
                    }
                    break;
                }
            }
        }
    }

    /// <summary>
    /// Get content type from file extension.
    /// </summary>
    private static string GetContentType(string filePath)
    {
        var ext = Path.GetExtension(filePath).ToLower();
        return ContentTypes.TryGetValue(ext, out var contentType) ? contentType : "application/octet-stream";
    }

    /// <summary>
    /// Get message type from content type.
    /// </summary>
    private static string GetMessageType(string contentType)
    {
        foreach (var (msgType, types) in SupportedTypes)
        {
            if (types.Contains(contentType))
            {
                return msgType;
            }
        }
        throw new Exception($"Unsupported content type: {contentType}");
    }

    /// <summary>
    /// Get maximum file size for a message type.
    /// </summary>
    private static long GetMaxSize(string messageType)
    {
        return messageType switch
        {
            "image" => ImageMaxSize,
            "document" => DocumentMaxSize,
            "video" => VideoMaxSize,
            "audio" => AudioMaxSize,
            _ => DocumentMaxSize
        };
    }

    /// <summary>
    /// Upload a file as an attachment.
    /// </summary>
    private static async Task<JsonElement> UploadAttachmentAsync(string filePath)
    {
        // Validate file exists
        if (!File.Exists(filePath))
        {
            throw new FileNotFoundException($"File not found: {filePath}");
        }

        var fileInfo = new FileInfo(filePath);
        var contentType = GetContentType(filePath);
        var messageType = GetMessageType(contentType);

        // Check file size
        var maxSize = GetMaxSize(messageType);
        if (fileInfo.Length > maxSize)
        {
            throw new Exception($"File too large: {fileInfo.Length} bytes (max {maxSize} bytes for {messageType})");
        }

        using var content = new MultipartFormDataContent();
        using var fileStream = File.OpenRead(filePath);
        using var fileContent = new StreamContent(fileStream);

        fileContent.Headers.ContentType = new MediaTypeHeaderValue(contentType);
        content.Add(fileContent, "file", Path.GetFileName(filePath));

        httpClient.DefaultRequestHeaders.Clear();
        httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", ApiToken);
        httpClient.DefaultRequestHeaders.Add("X-Tenant-ID", TenantId);

        var response = await httpClient.PostAsync($"{ApiUrl}/attachments", content);
        var responseBody = await response.Content.ReadAsStringAsync();

        if ((int)response.StatusCode == 413)
        {
            throw new Exception("File too large (server rejected)");
        }
        else if ((int)response.StatusCode == 415)
        {
            throw new Exception("Unsupported media type (server rejected)");
        }
        else if (!response.IsSuccessStatusCode)
        {
            throw new Exception($"API Error {(int)response.StatusCode}: {responseBody}");
        }

        return JsonSerializer.Deserialize<JsonElement>(responseBody);
    }

    /// <summary>
    /// Send a message with an attachment.
    /// </summary>
    private static async Task<JsonElement> SendMediaMessageAsync(string conversationId, string attachmentId, string messageType, string? caption = null)
    {
        var payload = new Dictionary<string, object>
        {
            ["conversation_id"] = conversationId,
            ["message_type"] = messageType,
            ["attachments"] = new[] { attachmentId }
        };

        if (!string.IsNullOrEmpty(caption))
        {
            payload["text"] = caption;
        }

        var json = JsonSerializer.Serialize(payload);
        var content = new StringContent(json, Encoding.UTF8, "application/json");

        httpClient.DefaultRequestHeaders.Clear();
        httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", ApiToken);
        httpClient.DefaultRequestHeaders.Add("X-Tenant-ID", TenantId);

        var response = await httpClient.PostAsync($"{ApiUrl}/messages", content);
        var responseBody = await response.Content.ReadAsStringAsync();

        if (!response.IsSuccessStatusCode)
        {
            throw new Exception($"API Error {(int)response.StatusCode}: {responseBody}");
        }

        return JsonSerializer.Deserialize<JsonElement>(responseBody);
    }

    /// <summary>
    /// Download an attachment by ID.
    /// </summary>
    private static async Task DownloadAttachmentAsync(string attachmentId, string outputPath)
    {
        httpClient.DefaultRequestHeaders.Clear();
        httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", ApiToken);
        httpClient.DefaultRequestHeaders.Add("X-Tenant-ID", TenantId);

        var response = await httpClient.GetAsync($"{ApiUrl}/attachments/{attachmentId}/download");

        if (!response.IsSuccessStatusCode)
        {
            var errorBody = await response.Content.ReadAsStringAsync();
            throw new Exception($"API Error {(int)response.StatusCode}: {errorBody}");
        }

        var fileBytes = await response.Content.ReadAsByteArrayAsync();
        await File.WriteAllBytesAsync(outputPath, fileBytes);
        Console.WriteLine($"Downloaded to: {outputPath}");
    }

    /// <summary>
    /// Upload and send an image message.
    /// </summary>
    private static async Task<JsonElement> SendImageAsync(string conversationId, string filePath, string? caption = null)
    {
        Console.WriteLine($"Uploading image: {filePath}");
        var attachment = await UploadAttachmentAsync(filePath);
        Console.WriteLine($"  Uploaded: {attachment.GetProperty("id").GetString()}");

        Console.WriteLine("Sending image message...");
        return await SendMediaMessageAsync(conversationId, attachment.GetProperty("id").GetString()!, "image", caption);
    }

    /// <summary>
    /// Upload and send a document message.
    /// </summary>
    private static async Task<JsonElement> SendDocumentAsync(string conversationId, string filePath, string? caption = null)
    {
        Console.WriteLine($"Uploading document: {filePath}");
        var attachment = await UploadAttachmentAsync(filePath);
        Console.WriteLine($"  Uploaded: {attachment.GetProperty("id").GetString()}");

        Console.WriteLine("Sending document message...");
        return await SendMediaMessageAsync(conversationId, attachment.GetProperty("id").GetString()!, "document", caption);
    }

    /// <summary>
    /// Upload and send a video message.
    /// </summary>
    private static async Task<JsonElement> SendVideoAsync(string conversationId, string filePath, string? caption = null)
    {
        Console.WriteLine($"Uploading video: {filePath}");
        var attachment = await UploadAttachmentAsync(filePath);
        Console.WriteLine($"  Uploaded: {attachment.GetProperty("id").GetString()}");

        Console.WriteLine("Sending video message...");
        return await SendMediaMessageAsync(conversationId, attachment.GetProperty("id").GetString()!, "video", caption);
    }

    /// <summary>
    /// Upload and send an audio message.
    /// </summary>
    private static async Task<JsonElement> SendAudioAsync(string conversationId, string filePath, string? caption = null)
    {
        Console.WriteLine($"Uploading audio: {filePath}");
        var attachment = await UploadAttachmentAsync(filePath);
        Console.WriteLine($"  Uploaded: {attachment.GetProperty("id").GetString()}");

        Console.WriteLine("Sending audio message...");
        return await SendMediaMessageAsync(conversationId, attachment.GetProperty("id").GetString()!, "audio", caption);
    }

    /// <summary>
    /// Demo: Upload a file and send it as a message.
    /// Automatically detects the appropriate message type.
    /// </summary>
    private static async Task DemoUploadAndSendAsync(string filePath)
    {
        var contentType = GetContentType(filePath);
        var messageType = GetMessageType(contentType);

        Console.WriteLine($"\n--- Sending {messageType} ---");
        Console.WriteLine($"File: {filePath}");
        Console.WriteLine($"Content-Type: {contentType}");

        var attachment = await UploadAttachmentAsync(filePath);
        Console.WriteLine("Attachment uploaded:");
        Console.WriteLine($"  ID: {attachment.GetProperty("id").GetString()}");
        Console.WriteLine($"  Filename: {attachment.GetProperty("filename").GetString()}");
        Console.WriteLine($"  Size: {attachment.GetProperty("file_size").GetInt64()} bytes");

        var message = await SendMediaMessageAsync(
            ConversationId,
            attachment.GetProperty("id").GetString()!,
            messageType,
            $"Here's a {messageType} file!"
        );

        Console.WriteLine("Message sent:");
        Console.WriteLine($"  ID: {message.GetProperty("id").GetString()}");
        Console.WriteLine($"  Status: {message.GetProperty("status").GetString()}");
        Console.WriteLine($"  Created at: {message.GetProperty("created_at").GetString()}");
    }

    /// <summary>
    /// Load environment variables from .env file if it exists.
    /// </summary>
    private static void LoadEnvFile()
    {
        var envPath = Path.Combine(Directory.GetCurrentDirectory(), ".env");
        if (!File.Exists(envPath)) return;

        foreach (var line in File.ReadAllLines(envPath))
        {
            var trimmedLine = line.Trim();
            if (string.IsNullOrEmpty(trimmedLine) || trimmedLine.StartsWith("#")) continue;

            var separatorIndex = trimmedLine.IndexOf('=');
            if (separatorIndex <= 0) continue;

            var key = trimmedLine.Substring(0, separatorIndex).Trim();
            var value = trimmedLine.Substring(separatorIndex + 1).Trim();

            if (string.IsNullOrEmpty(Environment.GetEnvironmentVariable(key)))
            {
                Environment.SetEnvironmentVariable(key, value);
            }
        }
    }
}
